src/common/error_code.cc

```diff
@@ -0,0 +1,135 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2017 Red Hat, Inc. <contact@redhat.com>
+ *
+ * Author: Adam C. Emerson <aemerson@redhat.com>
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License version
+ * 2.1, as published by the Free Software Foundation.  See file
+ * COPYING.
+ */
+
+#include <exception>
+
+#include "common/error_code.h"
+
+using boost::system::error_category;
+using boost::system::error_condition;
+using boost::system::generic_category;
+
+namespace ceph {
+
+// A category for error conditions particular to Ceph
+
+class ceph_error_category : public converting_category {
+public:
+  ceph_error_category(){}
+  const char* name() const noexcept override;
+  std::string message(int ev) const noexcept override;
+  using converting_category::equivalent;
+  bool equivalent(const boost::system::error_code& c,
+		  int ev) const noexcept override;
+  int from_code(int ev) const noexcept override;
+};
+
+const char* ceph_error_category::name() const noexcept {
+  return "ceph";
+}
+
+std::string ceph_error_category::message(int ev) const noexcept {
+  using namespace ::ceph::errc;
+
+  switch (ev) {
+  case 0:
+    return "No error";
+
+  case not_in_map:
+    return "Map does not contain requested entry.";
+  }
+
+  return "Unknown error.";
+}
+
+bool ceph_error_category::equivalent(const boost::system::error_code& c,
+				     int ev) const noexcept {
+  if (c.category() == generic_category() &&
+      c.value() == boost::system::errc::no_such_file_or_directory &&
+      ev == ceph::errc::not_in_map) {
+    // Blargh. A bunch of stuff returns ENOENT now, so just to be safe.
+    return true;
+  }
+  return false;
+}
+
+int ceph_error_category::from_code(int ev) const noexcept {
+  switch (ev) {
+  case 0:
+    return 0;
+  case ceph::errc::not_in_map:
+    // What we use now.
+    return -ENOENT;
+  }
+  return -EDOM;
```

Comment:
I'm just using it to mark 'This is an error we should expect not to get since we aren't using trig functions or anything else likely to produce it'.

Basically, if I see this it indicates an error code that needs to be converted and isn't.